[docs] Add themeable component guide

siriwatknp opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

**Material UI guide**: https://deploy-preview-37908--material-ui.netlify.app/material-ui/guides/themeable-component/
**Joy UI guide**: https://deploy-preview-37908--material-ui.netlify.app/joy-ui/guides/themeable-component/

All of the APIs are public but we've never published the docs. I think it's time to do it. Once the content is stable, I will add the same to Joy UI.

closes #32967

## Further improvements in v6 (breaking changes)

The case transforming logic can be removed completely to make the API more intuitive (what you pass in is what you get). Existing usage on Material UI and Joy UI needs to be updated accordingly.

```diff
// as an example, Joy UI Button.tsx
export const ButtonRoot = styled('button', {
  name: 'JoyButton',
+  slot: 'root',
-  slot: 'Root',
-  overridesResolver: (props, styles) => styles.root,
})(…)
```

---

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

siriwatknp commented:
@mnajdova I think the `styled` can be simplified a bit.

1. At this point, the `overridesResolver` is required to make the component themable but I think 99% of the custom component does not need custom `overridesResolver` (I think `overridesResolver` exists to make v4 -> v5 possible)

I think developers should provide just `name` and `slot`:

```js
const StatSlot = styled('div', {
  name: 'MuiStat',
  slot: 'Root',
})(…)
```

2. I think it will be better to use `slot` as a lower case or camel case. Basically whatever the slot value that you provide will be the key that the users will have to use.

Right now it's kinda confusing why I have to use `Root` in the slot but the users will use `root` in the theme.

I think both points won't be a breaking change.

cc @DiegoAndai

siriwatknp commented:
@samuelsycamore I put this in the how-to guide, not the customization. What do you think?

DiegoAndai reviewed (commented):
Nice example! Easy to follow and understand 😊🎉

1. Regarding `overridesResolver`, what do you propose? if nothing is provided, then the style override slot styles won't be added. Should we default to `styleOVerrides[slotName]`?
2. Regarding the `slot` casing, I agree we could use `camelCase` in this guide so it's the same as the key the users will use. Should we do that from now on in our own codebase?

siriwatknp commented on docs/data/material/guides/custom-component/custom-component.md:
```diff
@@ -0,0 +1,335 @@
+  fontWeight: 600,
+}));
+
+// …Do the same for the `value` and `unit` slots
```
Got it.

siriwatknp commented on docs/data/material/guides/custom-component/custom-component.md:
````diff
@@ -0,0 +1,335 @@
+
+Finally, you have to add the Stat component the theme types.
+
+```ts
````
Sure, I will add a comment to it.

siriwatknp commented:
> * Regarding `overridesResolver`, what do you propose? if nothing is provided, then the style override slot styles won't be added. Should we default to `styleOverrides[slotName]`?

I have the same thought. It should default to `styleOverrides[slotName]`. I don't have a use case to customize it in Joy UI so I expect the others will be the same.

> * Regarding the `slot` casing, I agree we could use `camelCase` in this guide so it's the same as the key the users will use. Should we do that from now on in our own codebase?

We should but we have to update the implementation because some code is checking the upper first letter like `Root`. If @mnajdova agrees with this, I will update it in this PR (no breaking change).

DiegoAndai commented:
I agree with both changes proposed by @siriwatknp in [the comment above](https://github.com/mui/material-ui/pull/37908#issuecomment-1631770292) 😊

mnajdova commented on docs/data/material/guides/custom-component/custom-component.md:
```diff
@@ -0,0 +1,335 @@
+}
+
+declare module '@mui/material/styles' {
+  interface ComponentsPropsList extends LabComponentsPropsList {}
```
```suggestion
  interface ComponentsPropsList extends LabComponentsPropsList {}
```

I wouldn't use the lab here, it would be confusing for people not familiar with it. I would propose we just add a note that if you are using the `@mui/lab` you should extend this interface.

mnajdova commented:
> At this point, the overridesResolver is required to make the component themable but I think 99% of the custom component does not need custom overridesResolver (I think overridesResolver exists to make v4 -> v5 possible)
> I think developers should provide just name and slot:
> 
> ```
> const StatSlot = styled('div', {
>   name: 'MuiStat',
>   slot: 'Root',
> })(…)
> ```

We can make it optional 👍 

> I think it will be better to use slot as a lower case or camel case. Basically whatever the slot value that you provide will be the key that the users will have to use.
> Right now it's kinda confusing why I have to use Root in the slot but the users will use root in the theme.

The `slot` is used for the display name - the name of the components, this is why initially it was uppercased. The overrides resolver is the thing that decides which key will be used in the style overrides. It could depend on the state and other factors, not just the slot name.

omerman commented:
This looks great! Im looking forward to work using these Apis 🙌

siriwatknp commented on packages/mui-system/src/createStyled.js:
```diff
@@ -219,7 +233,7 @@ export default function createStyled(input = {}) {
         let displayName;
         if (componentName) {
-          displayName = `${componentName}${componentSlot || ''}`;
+          displayName = `${componentName}${capitalize(componentSlot || '')}`;
```
@mnajdova based on your [comment](https://github.com/mui/material-ui/pull/37908#issuecomment-1637938923), I think the logic can be the other way around.

I believe that using slotName directly is easier to understand for developers. The `capitalize` I added is to preserve the previous behavior. In fact, it can be removed because it is for debugging and it won't be used in production.

We could remove it in the next major release?

mnajdova replied:
I think we can keep using `capitalize` it in the `displayName`

siriwatknp commented:
> Nice example! Easy to follow and understand 😊🎉
> 
> 1. Regarding `overridesResolver`, what do you propose? if nothing is provided, then the style override slot styles won't be added. Should we default to `styleOverrides[slotName]`?
> 2. Regarding the `slot` casing, I agree we could use `camelCase` in this guide so it's the same as the key the users will use. Should we do that from now on in our own codebase?

I have updated:
1. added default `overridesResolver`, so it will use `styleOverrides[slotName]` as you suggested.
2. Now, the `slot` casing can be anything, we should let developers own this part (from the code, it won't have an impact on the theming logic, meaning whatever slot name they define the styleOverrides will have to be that name)

mnajdova reviewed (approved):
The implementation looks good. Let's clarify in the TODO which version we are talking about, so that we can search easily. The copy-writing can be reviewed by @samuelsycamore

samuelsycamore commented:
I'm on it! Thanks for being patient! 🫡

samuelsycamore commented on docs/data/joy/guides/themeable-component/themeable-component.md:
```diff
@@ -0,0 +1,309 @@
+
+### 4. Support theme default props
+
+To let your users customize the default props of your component for different projects, you need to use the `useThemeProps` API.
```
```suggestion
To customize your component's default props for different projects, you need to use the `useThemeProps` API.
```

samuelsycamore commented on docs/data/joy/guides/themeable-component/themeable-component.md:
````diff
@@ -0,0 +1,309 @@
+  });
+```
+
+Then your users can customize the default props of your component like this:
````
```suggestion
Then you can customize the default props of your component like this:
```

samuelsycamore commented on docs/data/joy/guides/themeable-component/themeable-component.md:
```diff
@@ -0,0 +1,309 @@
+
+## TypeScript
+
+If you use TypeScript, you should create interfaces for the component props and ownerState.
```
```suggestion
If you use TypeScript, you must create interfaces for the component props and ownerState:
```

samuelsycamore commented on docs/data/joy/guides/themeable-component/themeable-component.md:
````diff
@@ -0,0 +1,309 @@
+});
+```
+
+Finally, you have to add the Stat component the theme types.
````
```suggestion
Finally, add the Stat component to the theme types.
```

samuelsycamore commented on docs/data/joy/guides/themeable-component/themeable-component.md:
````diff
@@ -0,0 +1,309 @@
+```
+
+:::success
+🍾 Congratulations! You have successfully built a themeable component!.
````
I would cut this (or at least remove the period)

```suggestion
```

samuelsycamore commented on docs/data/joy/guides/themeable-component/themeable-component.md:
```diff
@@ -0,0 +1,309 @@
+
+## Template
+
+You can use the following template to create your own component. It is the full code of the `Stat` component in this guide.
```
Reiterating the purpose of the template for anyone who skipped straight to it and might not be sure what they're looking at.

```suggestion
This template is the final product of the step-by-step guide above, demonstrating how to build a custom component that can be styled with the theme as if it was a built-in component.
```

samuelsycamore reviewed (commented):
I'm so sorry it took me so long to get back to this @siriwatknp !! 😭 I assume that most of my comments here apply to the Material UI doc as well.

siriwatknp commented on docs/data/joy/guides/themeable-component/themeable-component.md:
```diff
@@ -0,0 +1,309 @@
+
+### 1. Create the component slots
+
+The slots let your users customize each part of the component by targeting the slot's name in [theme's styleOverrides](/joy-ui/customization/themed-components/#theme-style-overrides).
```
Yes, that sounds better.

siriwatknp commented:
@samuelsycamore all fixed, including Material UI guide.

oliviertassinari commented:
I'm late to the party, but this change makes a lot of sense 👍
